src/kernel/bitcoinkernel.h

```diff
@@ -0,0 +1,1271 @@
+ * @param[in] block_tree_db_in_memory Set block tree db in memory.
+ */
+void kernel_chainstate_load_options_set_block_tree_db_in_memory(
+    kernel_ChainstateLoadOptions* chainstate_load_options,
+    bool block_tree_db_in_memory
+) BITCOINKERNEL_ARG_NONNULL(1);
+
+/**
+ * @brief Sets chainstate db in memory in the chainstate load options.
+ *
+ * @param[in] chainstate_load_options Non-null, created by @ref kernel_chainstate_load_options_create.
+ * @param[in] chainstate_db_in_memory Set chainstate db in memory.
+ */
+void kernel_chainstate_load_options_set_chainstate_db_in_memory(
+    kernel_ChainstateLoadOptions* chainstate_load_options,
+    bool chainstate_db_in_memory
+) BITCOINKERNEL_ARG_NONNULL(1);
+
+/**
+ * Destroy the chainstate load options
+ */
+void kernel_chainstate_load_options_destroy(kernel_ChainstateLoadOptions* chainstate_load_options);
+
+///@}
+
+/** @name ChainstateManager
+ * Functions for chainstate management.
+ */
+///@{
+
+/**
+ * @brief Create a chainstate manager. This is the main object for many
+ * validation tasks as well as for retrieving data from the chain. It is only
+ * valid for as long as the passed in context also remains in memory.
+ *
+ * @param[in] chainstate_manager_options Non-null, created by @ref kernel_chainstate_manager_options_create.
+ * @param[in] block_manager_options      Non-null, created by @ref kernel_block_manager_options_create.
+ * @param[in] context                    Non-null, the created chainstate manager will associate with this
+ *                                       kernel context for the duration of its lifetime. The same context
+ *                                       needs to be used for later interactions with the chainstate manager.
+ * @return                               The allocated chainstate manager, or null on error.
+ */
+kernel_ChainstateManager* BITCOINKERNEL_WARN_UNUSED_RESULT kernel_chainstate_manager_create(
+    const kernel_Context* context,
+    const kernel_ChainstateManagerOptions* chainstate_manager_options,
+    const kernel_BlockManagerOptions* block_manager_options
+) BITCOINKERNEL_ARG_NONNULL(1, 2, 3);
+
+/**
+ * @brief This function must be called to initialize the chainstate manager
+ * before doing validation tasks or interacting with its indexes.
+ *
+ * @param[in] context                 Non-null.
+ * @param[in] chainstate_load_options Non-null, created by @ref kernel_chainstate_load_options_create.
+ * @param[in] chainstate_manager      Non-null, will load the chainstate(s) and initialize indexes.
+ * @return                            True on success, false on error.
+ */
+bool BITCOINKERNEL_WARN_UNUSED_RESULT kernel_chainstate_manager_load_chainstate(
+    const kernel_Context* context,
+    const kernel_ChainstateLoadOptions* chainstate_load_options,
+    kernel_ChainstateManager* chainstate_manager
+) BITCOINKERNEL_ARG_NONNULL(1, 2, 3);
```

Comment:
When this function is called more than once, kernel crashes with an assertion error:

```
Assertion failed: (!m_ibd_chainstate), function InitializeChainstate, file validation.cpp, line 5655.
```

The solutions I see atm:
1. document that this function may only be called once for each chainman
1. add a field to `kernel_ChainstateManager*` to keep track of it being loaded already, return false and log an error
1. rework LoadChainstate logic to handle multiple calls gracefully
1. remove `kernel_chainstate_manager_load_chainstate` altogether and load chainstate during `kernel_chainstate_manager_create`.

I'm not sure if we really need a separate `*_load_chainstate` function, so if that's true, then option 4. would probably be preferable? I implemented it in https://github.com/TheCharlatan/bitcoin/compare/kernelApi...stickies-v:bitcoin:kernel/remove-load-chainstate, but in practice this probably should be a rebase instead of an extra commit. Options 1. and 2. seem easy enough to implement too, 3. is probably not the most sensible.

Reply:
> remove kernel_chainstate_manager_load_chainstate altogether and load chainstate during kernel_chainstate_manager_create.

I would like this a lot, but I wanted to keep a separate chainstate load function in case we ever land a "blocks-only read-only" chainstate manager, where we don't need to load any chainstates. I feel like making this a no-op could work, the simplest thing to do would probably be adding something along the lines of:
```
if (chainman.GetAll().size() > 0) return true;
```
to `kernel_chainstate_manager_load_chainstate`. But then again it would move us closer to a correct by construction setup if we'd do the constructing and loading all at once.